cuda/btx_cudainterval_callbacks.cpp

```diff
@@ -0,0 +1,589 @@
+  // these calls, so we need to maintain state between them.
+  // 
+  // Since 1 and 2 always happen one after another in a given thread, we just need
+  // to save the function name and launch ts keyed by {hostname, process, thread}.
+  // In (2) we can associate this information with the events, so when the results
+  // are finally available it can be looked up.
+  std::unordered_map<hpt_t, fn_ts_t> hpt_profiled_function_name_and_ts;
+  std::unordered_map<hp_event_t, fn_ts_t> hp_event_to_function_name_and_ts;
+
+  // Encapsulate complex context management state
+  CUDAContextManager context_manager;
+};
+
+using data_t = struct data_s;
+
+void set_entry_ts(data_t *data, hpt_t hpt, int64_t ts) {
+  data->entry_ts[hpt] = ts;
+}
+
+int64_t get_entry_ts(data_t *data, hpt_t hpt) {
+  return data->entry_ts.at(hpt);
+}
+
+std::string strip_event_class_name(const char *str) {
+  std::string temp(str);
+  std::smatch match;
+  std::regex_search(temp, match, std::regex(":(.*?)_?(?:entry|exit)?$"));
+
+  // The entire match is hold in the first item, sub_expressions 
+  // (parentheses delimited groups) are stored after.
+  assert( match.size() > 1 && "Event_class_name not matching regex.");
+
+  return match[1].str();
+}
+
+static void send_host_message(void *btx_handle, void *usr_data, int64_t ts,
+                              const char *event_class_name, const char *hostname, int64_t vpid,
+                              uint64_t vtid, bool err) {
+  std::string event_class_name_striped = strip_event_class_name(event_class_name);
+  const int64_t _start = get_entry_ts(static_cast<data_t *>(usr_data), {hostname, vpid, vtid});
+
+  btx_push_message_lttng_host(btx_handle, hostname, vpid, vtid, _start, BACKEND_CUDA,
+                              event_class_name_striped.c_str(), (ts - _start), err);
+}
+
+static void send_traffic_message(void *btx_handle, void *usr_data, int64_t ts,
+                                 const char *event_class_name, const char *hostname,
+                                 int64_t vpid, uint64_t vtid, bool err) {
+  std::string event_class_name_stripped = strip_event_class_name(event_class_name);
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t key = {hostname, vpid, vtid};
+  const int64_t _start = get_entry_ts(state, key);
+
+  if (!err) {
+    auto size = state->entry_traffic_size.at(key);
+    btx_push_message_lttng_traffic(btx_handle, hostname, vpid, vtid, _start, BACKEND_CUDA,
+                                   event_class_name_stripped.c_str(), size);
+  }
+}
+
+
+void btx_initialize_component(void **usr_data) { *usr_data = new data_t; }
+
+void btx_finalize_component(void *usr_data) {
+  delete static_cast<data_t *>(usr_data);
+}
+
+void entries_callback(void *btx_handle, void *usr_data, int64_t ts,
+                      const char *event_class_name, const char *hostname, int64_t vpid,
+                             uint64_t vtid) {
+  set_entry_ts(((data_t *)usr_data), {hostname, vpid, vtid}, ts);
```

Comment:
Yeah I think we can inline `set_entry_ts` :)

Reply:
Ya I think we can remove the helper, it is not really making code significantly shorter or easier to understand. The getter could be useful in having .at vs [] depending on if Debug or Release build, but that is maybe better done with a wrapper or subclass of std::map.